007-OFEP-flag-change-events.md

```diff
@@ -0,0 +1,66 @@
+## 007-OFEP-flag-change-events
+
+## State: DRAFTING
+
+Some flag SDKs support listening for flag value changes or general configuration changes ([launchdarkly](https://docs.launchdarkly.com/sdk/features/flag-changes), [cloudbees](https://docs.cloudbees.com/docs/cloudbees-feature-management/latest/reporting/configuration-fetched-handler), flagd). This can allow us to use an event-based paradigm for consuming flags. Client apps may use feature flags for characteristics that aren't specifically tied to a user-action, making imperative flag evaluation a less-than-ideal solution. A server application could listen to a flag that changes some operational behavior.
+
+Examples:
+
+- A web client's banner color could be updated when an associated flag is changed, without any user action.
+- A web service could subscribe to a flag that controls its log-level.
+
+## Design
+
+The provider interface and the OpenFeature client would be extended to have new functionality to register handlers for a particular flag value or the configuration in general. When an _application author_ registers a handler, the provider would react by appropriately registering a listener on it's SDK, polling it's REST API, etc. If the flag (or configuration in general) is updated, the provider would call the registered handler. The handler might be called with the flag key, or perhaps some other metadata pertaining to the configuration update. **Flag values would not be provided to the handler**, the handler would simply run, indicating the associated configuration had changed. The _application author_ would then perform a flag evaluation for the changed flag. This is consistent with the aforementioned flag systems, which do not evaluate the flags when they've changed. One reason is that no dynamic context can be reasonably provided in the case of events, since the event is driven by a change in the flag management system, not a user-action.
+
+Per-flag design:
+
+```ts
+const client = OpenFeature.getClient();
+
+// subscribe to changes in a string flag called "hex-color", and evaluate it when it's updated
+client.addHandler("hex-color", async (key) => {
```

Comment:
I like the look of this approach over the broad `configuration-change`.

May I also suggest the addition of a `client.addHandlerWithContext` method taking the `EvaluationContext` as an argument. In processes where the `EvaluationContext` is static for the full lifecycle of a process or client instance this will allow for an on change event to a specific flag. This would also reduce the need for `client.get{type}Value` calls at the top of each handler function.

```
client.addHandlerWithContext("hex-color", {"foo":"bar"}, (value) => {
     console.log(`hex-color value change: ${value}`);
})
```

Reply:
I don't follow this. What does the context provide here? It doesn't seem to be passed into your resulting function..

Reply:
I like the idea of providing a context to the handler, but I don't think it will enable us to get the values with a change event, we'd still need to pass the context and call the evaluation in the handler. That's because most SDKs themselves don't allow you to pass the context to their handler, and don't provide flag values when the handler fires:

>The event data consists only of the flag key. It does not contain a flag value, because in server-side SDKs, flags only have values when they are evaluated for a specific set of user properties. (https://docs.launchdarkly.com/sdk/features/flag-changes)

This is LD doc, but it's not unusual. Cloudbees also doesn't provider values with their event.

Reply:
I've left this out of my current implementation. For now we won't pass flag values to the handlers either. They will have to be evaluated.